src/Common/Coverage.h

```diff
@@ -0,0 +1,272 @@
+#pragma once
+
+#include <cstdio>
+
+#include <algorithm>
+#include <atomic>
+#include <filesystem>
+#include <functional>
+#include <iterator>
+#include <mutex>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <queue>
+#include <string_view>
+#include <type_traits>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <fmt/format.h>
+#include <Poco/Logger.h>
+
+#include "common/logger_useful.h"
+#include <common/types.h>
+
+#if defined(__ELF__) && !defined(__FreeBSD__)
+    #define NON_ELF_BUILD 0
+    #include <Common/SymbolIndex.h>
+    #include <Common/Dwarf.h>
+
+using SymbolIndex = DB::SymbolIndex;
+using SymbolIndexInstance = decltype(SymbolIndex::instance());
+using Dwarf = DB::Dwarf;
+
+#else
+    /// FreeBSD and Darwin do not have DWARF, so coverage build is explicitly disabled.
+    /// Fake classes are introduced to be able to build CH.
+
+    #define NON_ELF_BUILD 1
+    #if WITH_COVERAGE
+        #error "Coverage build does not work on FreeBSD and Darwin".
+    #endif
+
+struct SymbolIndexInstance
+{
+    struct SymbolPtr { std::string_view name; };
+    static constexpr SymbolPtr sym_ptr;
+
+    struct Ptr { constexpr const SymbolPtr * findSymbol(void*) const { return &sym_ptr; } }; //NOLINT
+    static constexpr Ptr ptr;
+
+    constexpr const Ptr * operator->() const { return &ptr; }
+};
+
+struct SymbolIndex { static constexpr SymbolIndexInstance instance() { return {}; } };
+
+struct Dwarf
+{
+    struct File { std::string toString() const { return {}; } }; //NOLINT
+    struct FileAndLine { size_t line; File file; };
+
+    constexpr FileAndLine findAddressForCoverageRuntime(uintptr_t) const { return {}; } //NOLINT
+};
+
+#endif
+
+namespace detail
+{
+using EdgeIndex = uint32_t;
+using Addr = void *;
+using Line = int;
+
+using Threads = std::vector<std::thread>;
+
+/**
+ * Simplified FreeThreadPool from Common/ThreadPool.h . Uses one external thread.
+ * Not intended for general use (some invariants broken).
+ *
+ * - Does not throw.
+ * - Does not use metrics.
+ * - Some operations are not thread-safe.
+ *
+ * Own implementation needed as Writer does not use most FreeThreadPool features (and we can save ~20 minutes by
+ * using this class instead of the former).
+ *
+ * Each test duration is longer than coverage test processing pipeline (converting and dumping to disk), so no
+ * more than 1 thread is needed.
+ */
+class TaskQueue
```

Comment:
How and why this is faster ? Why cant we create our thread pool with one thread ?

Reply:
We can't use internal CH pools as they would need to be excluded from instrumentation (due to runtime implementation).
TaskQueue _is_ a thread pool with one thread, the only difference is waiting on condition variable.
Motivation for that is clear in the comment:
>  Each test duration is longer than coverage test processing pipeline (converting and dumping to disk)